### PR TITLE
 *: Parameterize hotspot scheduling and adjust hot cache

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve? 
Improve stability and let hot region configurable.

### What is changed and how it works?

- fix the hot cache problem
- parameterize the schedule
- adjust the pendings

### Check List 

Tests 

 - Unit test
 - Manual test
   Command:  sysbench oltap_read_only (table size=2000000, threads=300)
   Before:
    image
  After:
   
  image

### CODE REVIEW

File: server/schedulers/hot_region.go [hunk 1]
```diff
@@ -228,7 +230,7 @@ func (h *hotScheduler) gcRegionPendings() {
 		empty := true
 		for ty, op := range pendings {
 			if op != nil && op.IsEnd() {
-				if time.Now().After(op.GetCreateTime().Add(minRegionScheduleInterval)) {
+				if time.Now().After(op.GetCreateTime().Add(h.conf.GetMaxZombieDuration())) {
```
Comment (rleungx) [inline server/schedulers/hot_region.go]: GetMaxZombieDuration？

File: tools/pd-ctl/pdctl/command/scheduler.go [hunk 2]
```diff
@@ -430,11 +430,29 @@ func NewConfigSchedulerCommand() *cobra.Command {
...
+		Use:   "balance-hot-region-scheduler",
+		Short: "show evict-leader-scheduler config",
+		Run:   listSchedulerConfigCommandFunc,
+	}
+	c.AddCommand(&cobra.Command{
+		Use:   "list",
+		Short: "list the config item",
+		Run:   listSchedulerConfigCommandFunc})
+	c.AddCommand(&cobra.Command{
+		Use:   "set",
```
Comment (rleungx) [inline tools/pd-ctl/pdctl/command/scheduler.go]: set <key> <value>?

File: server/schedulers/hot_region_config.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+	"github.com/gorilla/mux"
+	"github.com/pingcap/pd/v4/server/core"
+	"github.com/pingcap/pd/v4/server/schedule"
+	"github.com/pingcap/pd/v4/server/statistics"
+	"github.com/unrolled/render"
+)
+
+// params about hot region.
+var initHotRegionScheduleConfig = func() *hotRegionSchedulerConfig {
```
Comment (disksing) [inline server/schedulers/hot_region_config.go]: why not
```
func initHotRegionScheduleConfig() *hotRegionSchedulerConfig
```

File: server/schedulers/hot_region_config.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+	CountRankStepRatio    float64 `json:"count-rank-step-ratio"`
+	GreatDecRatio         float64 `json:"great-dec-ratio"`
+	MinorDecRatio         float64 `json:"minor-dec-ratio"`
+}
+
+func (conf *hotRegionSchedulerConfig) EncodeConfig() ([]byte, error) {
+	conf.RLock()
+	defer conf.RUnlock()
+	return schedule.EncodeConfig(conf)
+
```
Comment (disksing) [inline server/schedulers/hot_region_config.go]: rm empty line

File: server/schedulers/hot_region_config.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+	defer conf.RUnlock()
+	return conf.MinHotByteRate
+}
+
+func (conf *hotRegionSchedulerConfig) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	router := mux.NewRouter()
+	router.HandleFunc("/list", conf.handleGetConfig).Methods("GET")
+	router.HandleFunc("/config", conf.handleSetConfig).Methods("POST")
+	router.ServeHTTP(w, r)
+
```
Comment (disksing) [inline server/schedulers/hot_region_config.go]: empty line

File: server/schedulers/hot_region.go (hunk 1 above)
Comment (nolouch) [reply server/schedulers/hot_region.go]: Yes, it still influences the store, which should be consistent with the `pendingSum`.

File: server/statistics/hot_peer.go [hunk 6]
```diff
@@ -29,7 +29,7 @@ type HotPeerStat struct {
 	// HotDegree records the hot region update times
 	HotDegree int `json:"hot_degree"`
 	// AntiCount used to eliminate some noise when remove region in cache
-	AntiCount int
+	AntiCount int `json:"anti-count"`
```
Comment (disksing) [inline server/statistics/hot_peer.go]: anti_count

**Comment by nolouch** [comment]
ptal @disksing @rleungx

File: server/schedulers/hot_region.go [hunk 7]
```diff
@@ -513,7 +516,10 @@ func (bs *balanceSolver) solve() []*operator.Operator {
 	}
 
 	for i := 0; i < len(ops); i++ {
-		bs.sche.addPendingInfluence(ops[i], best.srcStoreID, best.dstStoreID, infls[i], bs.rwTy, bs.opTy)
+		// TODO: multiple operators need to be atomic.
+		if !bs.sche.addPendingInfluence(ops[i], best.srcStoreID, best.dstStoreID, infls[i], bs.rwTy, bs.opTy) {
+			return nil
```
Comment (lhy1024) [inline server/schedulers/hot_region.go]: Their region id may not be the same for different ops

File: server/schedulers/hot_region_config.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+}
+
+func (conf *hotRegionSchedulerConfig) handleGetConfig(w http.ResponseWriter, r *http.Request) {
+	conf.RLock()
+	defer conf.RUnlock()
+	rd := render.New(render.Options{IndentJSON: true})
+	rd.JSON(w, http.StatusOK, conf)
+}
+
+func (conf *hotRegionSchedulerConfig) handleSetConfig(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (nolouch) [reply server/schedulers/hot_region_config.go]: already add the test, check `scheduler_test.go`.

File: server/schedulers/hot_region.go (hunk 7 above)
Comment (nolouch) [reply server/schedulers/hot_region.go]: Yes, but now we only have one operator in each `Schedule()`. so I add a comment in here.

**Comment by nolouch** [comment]
Hi, @lhy1024  I update the test result in the issue comment, you can check it now.

File: server/schedulers/hot_region.go [hunk 9]
```diff
@@ -63,23 +72,7 @@ const (
 
 	hotRegionLimitFactor = 0.75
 
-	maxPeerNum = 1000
-
-	maxZombieDur time.Duration = statistics.StoreHeartBeatReportInterval * time.Second
-
 	minRegionScheduleInterval time.Duration = statistics.StoreHeartBeatReportInterval * time.Second
```
Comment (rleungx) [inline server/schedulers/hot_region.go]: It seems this variable can be moved in to test.

File: server/schedulers/hot_region.go [hunk 10]
```diff
@@ -98,9 +91,11 @@ type hotScheduler struct {
 	// temporary states but exported to API or metrics
 	stLoadInfos [resourceTypeLen]map[uint64]*storeLoadDetail
 	pendingSums [resourceTypeLen]map[uint64]Influence
+	// config of hot scheudler
```
Comment (rleungx) [inline server/schedulers/hot_region.go]:
```suggestion
	// config of hot scheduler
```

**Pull request merged**